src/hooks/useCustomLoaders.tsx

```diff
@@ -0,0 +1,58 @@
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import firebase from "firebase/app";
+
+import { CustomLoader, isCustomLoader } from "types/CustomLoader";
+
+export interface CustomLoadersState {
+  customLoaders: CustomLoader[];
+}
+
+export const initialValue: CustomLoadersState = {
+  customLoaders: [],
+};
+
+export const CustomLoadersContext = createContext<CustomLoadersState>(
+  initialValue
+);
+
+export const CustomLoadersProvider: React.FC = ({ children }) => {
+  const [customLoaders, setCustomLoaders] = useState<CustomLoader[]>([]);
+
+  // Fetch the loaders data on first load
+  useEffect(() => {
+    (async () => {
+      const loadersSnapshot = await firebase
+        .firestore()
+        .collection("loaders")
+        .get();
+
+      const loaders: CustomLoader[] = loadersSnapshot.docs
+        .map((d) => d.data())
+        .filter(isCustomLoader);
+
+      setCustomLoaders(loaders);
+    })();
+  }, []);
+
+  const providerData = useMemo(
+    () => ({
+      customLoaders,
+    }),
+    [customLoaders]
+  );
+
+  return (
+    <CustomLoadersContext.Provider value={providerData}>
+      {children}
+    </CustomLoadersContext.Provider>
+  );
+};
+
+export const useCustomLoaders = (): CustomLoadersState =>
+  useContext(CustomLoadersContext);
```

Comment:
[note] This implements a React context pattern in a clean way that encapsulates all of the logic, and ensures we can 'provide once, use many times' in an efficient way. This pattern will be useful for many things in the app (eg. reactions immediately comes to mind, assuming we don't fully rewrite it into redux)

(cc // @mike-lvov @denisdimitrov @sparkletown/engineering FYI)

Reply:
(backlogged to document properly in https://github.com/sparkletown/internal-sparkle-issues/issues/568)